public-api/viz split

alpeb opened the pull request:
These changes are grouped into 8 commits described and linked to as follows (I'll eventually followup with more commits addressing feedback).

###  1) [A couple more protobuf changes:](https://github.com/linkerd/linkerd2/pull/5560/commits/3b19e500ef9a49a9d29f6e80d19b6b8dfcf9348e)
    
- Moved `healthcheck.proto` back from viz to `proto/common` as it remains being used by the main `healthcheck.go` library (it was moved to viz by #5510).
- Extracted from `viz.proto` the IP-related types and put them in `/controller/gen/common/net` to be used by both the public and the viz APIs.

### 2) [Chart templates for new viz linkerd-metrics-api pod](https://github.com/linkerd/linkerd2/pull/5560/commits/02b188989e1304b8197169567b9f0451fa0ab7ea)

### 3) [Spin-off viz healthcheck:](https://github.com/linkerd/linkerd2/pull/5560/commits/729a14fa92138395aee909aa15e4cb2a67517672)
    
- Created `viz/pkg/healthcheck/healthcheck.go` that wraps the original `pkg/healthcheck/healthcheck.go` while adding the `vizNamespace` and `vizAPIClient` fields which were removed from the core `healthcheck`. That way the core healthcheck doesn't have any dependencies on viz, and viz' healthcheck can now be used to retrieve viz api clients.
- The core and viz healthcheck libs are now abstracted out via the new `healthcheck.Runner` interface.
- ~~Moved to viz the `linkerd-data-plane` checks because they rely on Prometheus, but they're currently commented-out pending the wiring up of `linkerd viz check --proxy`.~~ Refactored the data plane checks so they don't rely on calling `ListPods`
- ~~The core healthcheck unit tests dealing with the viz api have been removed and will be refactored later into a new set of viz healthcheck unit tests.~~
- The checks in `viz/cmd/check.go` have been moved to `viz/pkg/healthcheck/healthcheck.go` as well, so `check.go`'s sole responsibility is dealing with command business. This command also now retrieves its viz api client through viz' healthcheck.
- ~~The `getNamespaceOfExtensions()` function has been refactored into `viz.pkg.GetVizNamespace()` as is consumed from a few other places as well.~~
- `multicluster/cmd/check.go` also now relies on viz' api because it hits the `Gateways()` endpoint that depends on Prometheus. **In a followup, we should move that api into some MC component (the gateway itself?) so multicluster can be installed without viz**
- Ditto for `multicluster/cmd/gateways.go`

### 4) [Remove linkerd-controller dependency on Prometheus:](https://github.com/linkerd/linkerd2/pull/5560/commits/4b31b9a0ddb84628204a8a80e3a13cb9a0fb99d2)
    
- Removed the `global.prometheusUrl` config in the core values.yml.
- Removed `linkerd-controller` dependencies on Prometheus.
- Leave the Heartbeat's  `-prometheus` flag hard-coded temporarily. TO-DO: have it automatically discover viz and pull Prometheus' endpoint (#5352).

### 5) [Move observability gRPC from linkerd-controller to viz:](https://github.com/linkerd/linkerd2/pull/5560/commits/3a0919bd309810999ce5f187f6a62bf63000501f)
    
We continue removing the linkerd-controller dependencies on viz stuff, except for things required for the tap server, which can be tackled separately afterwards.
    
- Created a new gRPC server under `viz/metrics-api` moving prometheus-dependent functions out of the core gRPC server and into it (same thing for the accompaigning http server).
- Did the same for the `PublicAPIClient` (now called just `Client`) interface. The `VizAPIClient` interface disappears as it's enough to just rely on the viz `ApiClient` protobuf type.
- Moved the other files implementing the rest of the gRPC functions from `controller/api/public` to `viz/metrics-api` (`edge.go`, `stat_summary.go`, etc.).
- Also simplified some type names to avoid stuttering.

### 6)  [linkerd-metrics-api bootstrap files:](https://github.com/linkerd/linkerd2/pull/5560/commits/4845434f5a85d9ac40f9c66522a14dfd7081db3f)
    
Here we complete the work started in #5554 ("Chart templates for new viz linkerd-metrics-api pod") by providing the Dockerfile, `main.go` command and the `bin/` scripts for building the new `linkerd-metrics-api` pod under the `linkerd-viz` namespace.
    
At the same time, we strip out of the public-api's `main.go` file the prometheus parameters and other no longer relevant bits.
    
Finally, updated tests related to linkerd-metrics-api.

### 7) [linkerd-web updates:](https://github.com/linkerd/linkerd2/pull/5560/commits/161ac04ab397085a1c7cfb9c25ba89a146f9b8cc)
    
`linkerd-web` requires connecting with both the public-api and the viz api, so both addresses (and the viz namespace) are now provided as parameters to the container.
    
We're also updating some dependencies that still were pointing to the public-api and now should point to the viz api.

### 8) [CLI updates and other minor things:](https://github.com/linkerd/linkerd2/pull/5560/commits/f8371c4d90bde4597e203c023a1822edd5c61e29)
    
Changes to command files under `cli/cmd`:
- Updated `endpoints.go` according to new API interface name.
- Updated `version.go`, `dashboard` and `uninstall.go` to pull the viz namespace dynamically.
    
Changes to command files under `viz/cmd`:
- `edges.go`, `routes.go`, `stat.go` and `top.go`: point to dependencies that were moved from public-api to viz.
    
Other changes to have tests pass:
- Added `metrics-api` to list of docker images to build in actions workflows.
- In `bin/fmt` exclude protobuf generated files instead of entire directories because directories could contain both generated and non-generated code (case in point: `viz/metrics-api`).
- Update Helm readme files.

Closes #5328

kleimkuhler commented on viz/metrics-api/client/client.go:
```diff
@@ -0,0 +1,200 @@
+	return &msg, err
+}
+
+func (c *grpcOverHTTPClient) Tap(ctx context.Context, req *pb.TapRequest, _ ...grpc.CallOption) (pb.Api_TapClient, error) {
+	return nil, status.Error(codes.Unimplemented, "Tap is deprecated in public API, use tap APIServer")
+}
+
+func (c *grpcOverHTTPClient) TapByResource(ctx context.Context, req *pb.TapByResourceRequest, _ ...grpc.CallOption) (pb.Api_TapByResourceClient, error) {
+	return nil, status.Error(codes.Unimplemented, "Tap is deprecated in public API, use tap APIServer")
+}
```
We should remove the "public" part from these. Maybe even we can get rid of them now?

alpeb replied:
Indeed. I've just pushed the changes related to that removal :+1:

kleimkuhler commented on viz/metrics-api/client/client.go:
```diff
@@ -0,0 +1,200 @@
+
+// NewInternalClient creates a new Viz API client intended to run inside a
+// Kubernetes cluster.
+func NewInternalClient(namespace string, kubeAPIHost string) (pb.ApiClient, error) {
```
May be worth nothing this and `NewExternalClient` don't have that dance I commented above in public API with one function wrapping another but not for any particular reason.

kleimkuhler reviewed (commented):
Still have some more reviewing testing to do, but leaving first round of comments.

alpeb commented on controller/api/public/grpc_server.go:
```diff
@@ -3,200 +3,55 @@ package public
-		ignoredNamespaces:   ignoredNamespaces,
 	}
 
 	pb.RegisterApiServer(prometheus.NewGrpcServer(), grpcServer)
```
If you mean the `prometheusAPI` field, that still is used in the viz gRPC file (`viz/metrics-api/grpc_server.go`). The public-api should have no dependencies on prometheus.

adleong replied:
I think this is for instrumenting the server and exposing controller metrics (which can be scraped by prometheus)

kleimkuhler replied:
@alpeb Nope I'm referring to the `pb.RegisterApiServer(prometheus.NewGrpcServer(), grpcServer)` line 43.

And if that is already taken care of by the admin server then maybe this can be removed then?

alpeb replied:
Ok then that does look like something about the prom metrics that are being served. But if I comment that out, the metrics served look the same (go and http related metrics). I honestly have no idea what that does, and it's not documented in those functions either :man_shrugging:

adleong commented on controller/api/public/grpc_server.go:
```diff
@@ -3,200 +3,55 @@ package public
 }
 
-func (s *grpcServer) SelfCheck(ctx context.Context, in *healthcheckPb.SelfCheckRequest) (*healthcheckPb.SelfCheckResponse, error) {
+// TODO: Reenable this check just for checking the control plane can
```
In order to make this work without taking a dependency on linkerd-viz, we'd need to introduce a new API or move SelfCheck to a shared location.  To be honest, I'm not sure if it's even worth it.  Especially since this process doesn't even need to talk to the k8s API.  Perhaps a better check would be to look for errors in the destination or identity controller logs or kubernetes event streams?

adleong replied:
But I think we can get rid of this TODO

alpeb commented on controller/api/util/api_utils.go:
```diff
@@ -5,13 +5,12 @@ import (
 
+	netPb "github.com/linkerd/linkerd2/controller/gen/common/net"
 	"github.com/linkerd/linkerd2/pkg/k8s"
 	pb "github.com/linkerd/linkerd2/viz/metrics-api/gen/viz"
```
Agreed about Tap. OTOH, that `Resource` type is pretty generic, so we can also think about moving it into `proto/common`.

adleong replied:
The end goal should definitely be to get rid of all references to viz in the main control plane.  Probably we need a new internal `Resource` struct (or perhaps there's something in API-machinery we can use) and these `BuildResource` functions should return that instead of a protobuf struct.  Then the Tap CLI will need to be able to convert from that resource type into a protobuf resource struct.

adleong replied:
@alpeb mind opening up a tracking issue for this if one doesn't already exist?

alpeb replied:
I've created #5589 :+1:

adleong commented on multicluster/cmd/check.go:
```diff
@@ -16,6 +16,7 @@ import (
 	"github.com/linkerd/linkerd2/pkg/servicemirror"
 	"github.com/linkerd/linkerd2/pkg/tls"
 	public "github.com/linkerd/linkerd2/viz/metrics-api/gen/viz"
+	vizHealthCheck "github.com/linkerd/linkerd2/viz/pkg/healthcheck"
```
Why does the mutlicluster check command use a viz healthchecker?

alpeb replied:
Because the `Gateways()` API hits prometheus to retrieve the liveness, latency and service count data.

adleong replied:
Ahh, but I think this means that `multicluster check` will fail if the linkerd-viz extension isn't installed?  Ideally the `multicluster check` would only perform the gateways check if the linkerd-viz extension is installed, and skip that particular check otherwise.

adleong replied:
Here's the error I get when running `mc check` without first installing viz:

```
 ✗ l mc check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x23654f2]

goroutine 1 [running]:
github.com/linkerd/linkerd2/viz/pkg/healthcheck.(*HealthChecker).vizCategory.func1(0x2aacfe0, 0xc00035ea20, 0x6fc23ac00, 0x2aacfe0)
	/linkerd-build/viz/pkg/healthcheck/healthcheck.go:65 +0x72
github.com/linkerd/linkerd2/pkg/healthcheck.(*HealthChecker).runCheck(0xc00053e840, 0x27b5a0b, 0xb, 0xc00071b9f0, 0xc000890030, 0x100c000)
	/linkerd-build/pkg/healthcheck/healthcheck.go:1506 +0x14c
github.com/linkerd/linkerd2/pkg/healthcheck.(*HealthChecker).RunChecks(0xc00053e840, 0xc000890030, 0xc000890030)
	/linkerd-build/pkg/healthcheck/healthcheck.go:1470 +0x298
github.com/linkerd/linkerd2/viz/pkg/healthcheck.(*HealthChecker).RunChecks(0xc00085ff80, 0xc000890030, 0x4)
	/linkerd-build/viz/pkg/healthcheck/healthcheck.go:54 +0x38
github.com/linkerd/linkerd2/pkg/healthcheck.runChecksTable(0x2a655a0, 0xc00000e018, 0x2a64000, 0xc00085ff80, 0x0)
	/linkerd-build/pkg/healthcheck/healthcheck_output.go:82 +0x14e
github.com/linkerd/linkerd2/pkg/healthcheck.RunChecks(0x2a655a0, 0xc00000e018, 0x2a655a0, 0xc00000e020, 0x2a64000, 0xc00085ff80, 0x27add7c, 0x5, 0x0)
	/linkerd-build/pkg/healthcheck/healthcheck_output.go:36 +0x68
github.com/linkerd/linkerd2/multicluster/cmd.configureAndRunChecks(0x2a655a0, 0xc00000e018, 0x2a655a0, 0xc00000e020, 0xc0004a96a0, 0xc00029eb00, 0x0)
	/linkerd-build/multicluster/cmd/check.go:125 +0x580
github.com/linkerd/linkerd2/multicluster/cmd.newCmdCheck.func1(0xc000927b80, 0x3938b60, 0x0, 0x0, 0x0, 0x0)
	/linkerd-build/multicluster/cmd/check.go:92 +0x219
github.com/spf13/cobra.(*Command).execute(0xc000927b80, 0x3938b60, 0x0, 0x0, 0xc000927b80, 0x3938b60)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0x38ec200, 0x0, 0x24a9a20, 0xc00005c0b8)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/linkerd-build/cli/main.go:10 +0x2d
```

alpeb replied:
Done, as describe below :+1:

alpeb commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1326,15 +1294,40 @@ func (hc *HealthChecker) allCategories() []Category {
 				},
 			},
 		},
+		{
+			id: LinkerdControlPlaneVersionChecks,
+			checkers: []Checker{
+				{
+					description: "control plane is up-to-date",
+					hintAnchor:  "l5d-version-control",
+					warning:     true,
+					check: func(context.Context) error {
+						return hc.latestVersions.Match(hc.serverVersion)
+					},
+				},
+				{
+					description: "control plane and cli versions match",
+					hintAnchor:  "l5d-version-control",
+					warning:     true,
+					check: func(context.Context) error {
+						if hc.serverVersion != version.Version {
+							return fmt.Errorf("control plane running %s but cli running %s", hc.serverVersion, version.Version)
+						}
+						return nil
+					},
+				},
+			},
+		},
```
My mistake :grimacing:

alpeb commented on multicluster/cmd/gateways.go:
```diff
@@ -41,7 +41,14 @@ func newGatewaysCommand() *cobra.Command {
+			ctx := cmd.Context()
+
+			vizNs, err := k8sAPI.GetNamespaceWithExtensionLabel(ctx, "linkerd-viz")
+			if err != nil {
+				return fmt.Errorf("make sure the linkerd-viz extension is installed, using 'linkerd viz install' (%s)", err)
+			}
```
#5328 is already solved here, in that the viz api location is discovered. Another thing is removing any mc dependency on viz, which I'm not sure how will that be done. We could move the `Gateways()` API out of viz and into somewher in the multicluster namespace, but that will still have to rely on some prometheus instance. Unless we stop caring about time-depending metrics like latency distributions, and only care about liveness and service count for example, and the service mirror controller should give us that.

adleong replied:
Yup, I think what's here is the right solution to #5328.

kleimkuhler replied:
Yep I was referring to what @alpeb mentioned about the multicluster dependency on Viz.

It doesn't need to be addressed as part of this, but I think if we remove that dependency and therefore lose the time-depending metrics like latency distributions, it's worth considering a little more.

I'd be curious how many users start doing a core and multicluster install, but wouldn't do a Viz install if they didn't need to.

kleimkuhler replied:
Oh it looks like this has been addressed! I had not read the comments down below yet 😬.

alpeb commented on multicluster/cmd/gateways.go:
```diff
@@ -41,7 +41,14 @@ func newGatewaysCommand() *cobra.Command {
-			client, err := public.NewExternalClient(cmd.Context(), controlPlaneNamespace, k8sAPI)
+			ctx := cmd.Context()
+
+			vizNs, err := k8sAPI.GetNamespaceWithExtensionLabel(ctx, "linkerd-viz")
```
You mean `vizChartName`? We'd have to introduce a new one in case the chart name changes. Also I see many commands in viz are still using the string instead of a const, so I'd rather leave this for general cleanup aftewards.

alpeb replied:
I see it was me that added some of those strings after all. I'll create the constant then.

adleong commented:
I'm getting a panic when trying to use viz commands before installing the viz extension:

```
l viz stat deploy -n linkerd
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x23654f2]

goroutine 1 [running]:
github.com/linkerd/linkerd2/viz/pkg/healthcheck.(*HealthChecker).vizCategory.func1(0x2aacfe0, 0xc0003bfe00, 0x6fc23ac00, 0x2aacfe0)
	/linkerd-build/viz/pkg/healthcheck/healthcheck.go:65 +0x72
github.com/linkerd/linkerd2/pkg/healthcheck.(*HealthChecker).runCheck(0xc0002c6600, 0x27b5a0b, 0xb, 0xc000a878e8, 0x28986d8, 0x0)
	/linkerd-build/pkg/healthcheck/healthcheck.go:1506 +0x14c
github.com/linkerd/linkerd2/pkg/healthcheck.(*HealthChecker).RunChecks(0xc0002c6600, 0x28986d8, 0x2)
	/linkerd-build/pkg/healthcheck/healthcheck.go:1470 +0x298
github.com/linkerd/linkerd2/viz/pkg/healthcheck.(*HealthChecker).RunChecks(...)
	/linkerd-build/viz/pkg/healthcheck/healthcheck.go:54
github.com/linkerd/linkerd2/viz/pkg/api.CheckClientOrRetryOrExit(0x27b05cd, 0x7, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/linkerd-build/viz/pkg/api/api.go:37 +0xff
github.com/linkerd/linkerd2/viz/pkg/api.CheckClientOrExit(...)
	/linkerd-build/viz/pkg/api/api.go:18
github.com/linkerd/linkerd2/viz/cmd.NewCmdStat.func1(0xc0008b1340, 0xc0008e85a0, 0x1, 0x3, 0x0, 0x0)
	/linkerd-build/viz/cmd/stat.go:183 +0x21d
github.com/spf13/cobra.(*Command).execute(0xc0008b1340, 0xc0008e8570, 0x3, 0x3, 0xc0008b1340, 0xc0008e8570)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x453
github.com/spf13/cobra.(*Command).ExecuteC(0x38ec200, 0x0, 0x24a9a20, 0xc000182058)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x349
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/linkerd-build/cli/main.go:10 +0x2d
```

alpeb commented:
Most of the feedback has been addressed (still need to see if I can get rid of the tap functions in the public-api, and to have `linkerd check` skip the `Gateways()` if viz is absent):

- Fixed panic when issuing viz commands when the extension is not
  installed.
- Removed unnecessary wrapping functions `NewInternalPublicClient` and `NewExternalPublicClient`.
- Created new constant in `viz/cmd/root` holding "linkerd-viz".
- Moved `LinkerdControlPlaneVersionChecks` in `healthcheck.go` to its original location in that same file.
- Removed commented out `SelfCheck` in the public-api's `grpc_server.go`

alpeb commented:
In the last push I rolled back `multicluster/cmd/check.go` changes so that we keep on using
the public-api client, instead of viz'. If the viz extension is not
detected then the "all gateway mirrors are healthy" check will just
issue a warning as show below. Otherwise, a viz client is instantiated and
`Gateways()` is called on it.

```
$ linkerdf cli mc check
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-multicluster
--------------------
√ Link CRD exists
√ Link resources are valid
        * target
√ remote cluster access credentials are valid
        * target
√ clusters share trust anchors
        * target
√ service mirror controller has required permissions
        * target
√ service mirror controllers are running
        * target
× all gateway mirrors are healthy
        failed to fetch gateway metrics: could not find the linkerd-viz extension
    see https://linkerd.io/checks/#l5d-multicluster-gateways-endpoints for hints
√ all mirror services have endpoints
√ all mirror services are part of a Link

Status check results are ×
```

kleimkuhler commented on viz/pkg/healthcheck/healthcheck.go:
```diff
@@ -0,0 +1,215 @@
+				return hc.CheckCertAndAnchorsExpiringSoon(cert)
+			}))
+
+	checkers = append(checkers,
+		*healthcheck.NewChecker("tap API service is running").
+			WithHintAnchor("l5d-tap-api").
+			Warning().
+			WithCheck(func(ctx context.Context) error {
+				return hc.CheckAPIService(ctx, linkerdTapAPIServiceName)
+			}))
```
It may be worth adding a retry to this checker now. When running `linkerd viz check` immediately after install, it produces the following warning:

```
‼ tap API service is running
    MissingEndpoints: endpoints for service/linkerd-tap in "linkerd-viz" have no addresses with port name "apiserver"
    see https://linkerd.io/checks/#l5d-tap-api for hints
```


Once you run `check` again though and the tap service is up and running it's fine.

alpeb replied:
Makes sense. I've just added that into my latest push.

kleimkuhler reviewed (commented):
Sorry there are a few merge conflicts after the tap-injector merge!

kleimkuhler reviewed (approved):
Nice! 🏁

adleong reviewed (commented):
I think the list of images in `bin/docker-push` needs to be updated.

adleong commented on multicluster/cmd/check.go:
```diff
@@ -464,12 +466,24 @@ func (hc *healthChecker) checkIfGatewayMirrorsHaveEndpoints(ctx context.Context)
+
+		vizNs, err := hc.linkerdHC.KubeAPIClient().GetNamespaceWithExtensionLabel(ctx, vizCmd.ExtensionName)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("failed to fetch gateway metrics: %s", err))
```
I don't have a working multicluster setup right now so I can't test this, but just reading this it looks like it would return an error and fail the check if the viz extension is not installed.  I'd expect that if the viz extension is not installed that this would return a `healthcheck.SkipError` to skip this check.

alpeb replied:
Yeah, currently it would show an error as shown [above](https://github.com/linkerd/linkerd2/pull/5560#issuecomment-764206298). We could also make it a warning and still have the overall `linkerd mc check` succeed, which might be more explicit. WDYT?

adleong replied:
I would return `SkipError`.  There's no hard dependency on viz from multicluster so if viz is unavailable, that shouldn't be considered a problem.  It just means that if viz *is* available, then we can do some extra validation of the stats.

alpeb commented:
In the last push, `linkerd mc check` doesn't error when viz is not available. I'm also properly setting the log in `multicluster/cmd/root.go` so that it properly displays messages when `--verbose` is used:

```bash
$ bin/go-run cli mc check 
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
√ can initialize the client
√ can query the control plane API

linkerd-multicluster
--------------------
√ Link CRD exists
√ Link resources are valid
        * target
√ remote cluster access credentials are valid
        * target
√ clusters share trust anchors
        * target
√ service mirror controller has required permissions
        * target
√ service mirror controllers are running
        * target
√ all mirror services have endpoints
√ all mirror services are part of a Link

Status check results are √



$ bin/go-run cli mc check --verbose
kubernetes-api
--------------
√ can initialize the client
√ can query the Kubernetes API

linkerd-existence
-----------------
√ 'linkerd-config' config map exists
√ heartbeat ServiceAccount exist
√ control plane replica sets are ready
√ no unschedulable pods
√ controller pod is running
DEBU[0000] Starting port forward to https://0.0.0.0:43271/api/v1/namespaces/linkerd/pods/linkerd-controller-85b69dcc46-84pdb/portforward?timeout=30s 34587:8085 
DEBU[0000] Port forward initialised                     
DEBU[0000] Expecting API to be served over [http://localhost:34587/api/v1/] 
√ can initialize the client
DEBU[0000] Making gRPC-over-HTTP call to [http://localhost:34587/api/v1/Version] [] 
DEBU[0000] Response from [http://localhost:34587/api/v1/Version] had headers: map[Content-Length:[56] Content-Type:[application/octet-stream] Date:[Thu, 21 Jan 2021 22:38:10 GMT]] 
DEBU[0000] gRPC-over-HTTP call returned status [200 OK] and content length [56] 
√ can query the control plane API

linkerd-multicluster
--------------------
√ Link CRD exists
√ Link resources are valid
        * target
√ remote cluster access credentials are valid
        * target
√ clusters share trust anchors
        * target
√ service mirror controller has required permissions
        * target
√ service mirror controllers are running
        * target
DEBU[0000] Skipping check: all gateway mirrors are healthy. Reason: failed to fetch gateway metrics 
√ all mirror services have endpoints
√ all mirror services are part of a Link

Status check results are √
```

adleong reviewed (approved):
:taco::taco::taco:
